File: 2023/03/t.py
cal = open("input.txt", "r", encoding="utf-8").read().split("\n")

for x in range(len(cal)):
    cal[x] = "." + cal[x] + "."

cal.insert(0, "." * len(cal[0]))
cal.append("." * len(cal[0]))

digits = "0123456789"
digitsPoint = digits + "."


def TestaNum(lin, inic, fim):
    for x in range(lin - 1, lin + 2):
        for y in range(inic - 1, fim + 2):
            if not cal[x][y] in digitsPoint:
                return int(cal[lin][inic : fim + 1])
    return 0


s1 = s2 = 0
for x in range(1, len(cal) - 1):
    y = 1
    while y <= len(cal[0]) - 1:
        if cal[x][y] in digits:
            inic = y
            y += 1
            while cal[x][y] in digits:
                y += 1
            fim = y - 1
            s1 += TestaNum(x, inic, fim)
        else:
            y += 1


def buscaNum(lin, col, v):
    if cal[lin][col] in digits:
        while cal[lin][col - 1] in digits:
            col -= 1
        s = ""
        while cal[lin][col] in digits:
            s += cal[lin][col]
            col += 1

        v.append(int(s))
        return


def TestaMult(lin, col):
    v = []
    for x in range(lin - 1, lin + 2):
        if cal[x][col] in digits:
            buscaNum(x, col, v)
        else:
            buscaNum(x, col - 1, v)
            buscaNum(x, col + 1, v)

    if len(v) > 1:
        return v[0] * v[1]
    else:
        return 0


for x in range(1, len(cal) - 1):
    for y in range(1, len(cal[0])):
        if cal[x][y] == "*":
            s2 += TestaMult(x, y)

print("Round 1:", s1)  # 546312
print("Round 2:", s2)  # 87449461
# input2.txt return 4361 and 467835, as defined in the site

File: 2023/03/t.cs
var linesFile = File.ReadAllLines("input.txt");

var cal = new string[linesFile.Length + 2];

var x = 1;
foreach (string line in linesFile)
{
    cal[x++] = "." + line + ".";
}
cal[0] = new String('.', cal[1].Length);
cal[x] = cal[0];

var digits = "0123456789";
var digitsPoint = digits + ".";

int s1 = 0;
int s2 = 0;

for (x = 1; x < cal.Length - 1; x++)
{
    var y = 1;
    while (y <= cal[0].Length - 1)
    {
        if (digits.Contains(cal[x][y]))
        {
            var inic = y;
            y++;
            while (digits.Contains(cal[x][y]))
            {
                y++;
            }
            var fim = y - 1;
            s1 += TestaNum(x, inic, fim);
        }
        else
        {
            y += 1;
        }
    }
}

for (x = 1; x < cal.Length - 1; x++)
{
    for (var y = 1; y <= cal[0].Length - 1; y++)
    {
        if (cal[x][y] == '*')
        {
            s2 += TestaMult(x, y);
        }
    }
}

Console.WriteLine($"Round 1: {s1}"); // 546312
Console.WriteLine($"Round 2: {s2}"); // 87449461

int TestaNum(int lin, int inic, int fim)
{
    for (var x = lin - 1; x <= lin + 1; x++)
    {
        for (var y = inic - 1; y <= fim + 1; y++)
        {
            if (!digitsPoint.Contains(cal[x][y]))
            {
                return Int32.Parse(cal[lin].Substring(inic, fim - inic + 1));
            }
        }
    }
    return 0;
}

void buscaNum(int lin, int col, List<int> list)
{
    if (digits.Contains(cal[lin][col]))
    {
        while (digits.Contains(cal[lin][col - 1]))
        {
            col--;
        }
        var s = "";
        while (digits.Contains(cal[lin][col]))
        {
            s += cal[lin][col];
            col++;
        }
        list.Add(Int32.Parse(s));
    }
}

int TestaMult(int lin, int col)
{
    List<int> list = new List<int>();
    for (var x = lin - 1; x <= lin + 1; x++)
    {
        if (digits.Contains(cal[x][col]))
        {
            buscaNum(x, col, list);
        }
        else
        {
            buscaNum(x, col - 1, list);
            buscaNum(x, col + 1, list);
        }
    }
    if (list.Count > 1)
    {
        return list[0] * list[1];
    }
    else
    {
        return 0;
    }
}